### DISCUSSION TITLE
Approach for debugging individual branches in dynamic targets

### DISCUSSION THREAD

**Discussion by lindsayplatt** [ropensci/targets#754] [Help] [answered]
Many of my workflows use dynamic branching for downloading or processing large data files. Sometimes each branch can take 10+ minutes to build. Sometimes the first 200+ branches run just fine and then the 201st throws an error. When this happens, I need to 

1. **Figure out what inputs went into this branch** to see what might be causing the error and fix the code.
2. **Re-build just this branch** to test that it will work (running `tar_make()` for just the branch target name does not work)
3. Ideally, **kick off the build for the dynamic target from this branch forward** to make sure no other branches fail and need fixes (I know that this is a bit risky BUT it happens all the time where I fix one branch, rebuild from the start since the function changed only to find that one later down the line failed and I have to start again). 

I haven't found a great way to do any of these steps. I sometimes decide to go back to the function and add `message()` calls to print out which inputs are being used for the current branch, but this clogs up the console and no longer allows the first 200 to be skipped since I changed the function. When I run `tar_meta()` on the specific branch, I see the `command` hash which I think has the information I am looking for but I don't know what to do with that hash.

As for skipping the ones that already worked and just running the current branch or ones from that branch forward, I keep thinking that something with `tarchetypes::tar_cue_skip()` should work but haven't figured out what yet.

**Comment by wlandau** [comment, marked as answer]
For (1) and (2), you might like https://books.ropensci.org/targets/debugging.html#workspaces. With `tar_option_set(workspace_on_error = TRUE)`, `targets` logs info about failed targets which allows you to load the dependencies, packages, and RNG seed in an interactive session. Works with branches. For (3), you can set `tar_option_set(error = "continue")` so the pipeline keeps going on error.

**Reply by lindsayplatt** [reply]
I can't believe I haven't noticed the workspaces functions before. Thanks! That will definitely help with debugging and fixing a branch that is throwing an error.

For (3), setting `error = "continue"` is not exactly what I am looking for. In particular, I am thinking that I completed (1) and (2), so I found and then edited code to fix the error that originally cropped up. Now the function code is different which would cause a rebuild of all the branches that had already successfully built. While I know that you do need to rebuild everything to make sure that your newest changes do still work with previously successful branches, a common pattern I fall into is finding a new edge case for the 250th branch, which requires the code to change again and for me to rebuild from the start once again. This wouldn't be a problem if each branch took a few seconds, but at 10+ minutes, waiting for the ones that _did aready succeed_ to build again while I am still verifying that all the edge cases are caught eats up a lot of time. I am hoping there is a way to temporarily skip branches that already succeeded even if the function changed slightly? I _think_ using `tarchetypes::tar_cue_skip()` will do this?

**Reply by wlandau** [reply]
To skip targets that already ran successfully at least once, you can write `tar_target(cue = tar_cue(mode = "never"))`. `tar_cue_skip()` is for when you have a more nuanced skipping condition in mind.

**Reply by lindsayplatt** [reply]
Just tested this with my long-running dynamic branch target and it worked very well. It was great to add workspaces to this target and not have any rebuilds happen, just recorded the workspaces for ones that were already complete and moved along. This will definitely be a timesaver!
